common/modulecontext/builder.go

```diff
@@ -0,0 +1,254 @@
+}
+
+func buildConfigOrSecrets[R cf.Role](ctx context.Context, b Builder, manager cf.Manager[R], items []configOrSecretItem) error {
+	for _, item := range items {
+		switch item := item.(type) {
+		case refValuePair:
+			isData, value, data, err := item.resolver.Resolve()
+			if err != nil {
+				return err
+			}
+			if isData {
+				if err := manager.SetData(ctx, cf.Ref(item.ref), data); err != nil {
+					return err
+				}
+			} else {
+				if err := manager.Set(ctx, cf.Ref(item.ref), value); err != nil {
+					return err
+				}
+			}
+		case configManager[R]:
+			list, err := item.manager.List(ctx)
+			if err != nil {
+				return err
+			}
+			for _, e := range list {
+				if m, isModuleSpecific := e.Module.Get(); isModuleSpecific && b.moduleName != m {
+					continue
+				}
+				data, err := item.manager.GetData(ctx, e.Ref)
+				if err != nil {
+					return err
+				}
+				if err := manager.SetData(ctx, e.Ref, data); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (b Builder) AddDSNsFromEnvarsForModule(module *schema.Module) Builder {
+	// remove in favor of a non-envar approach once it is available
+	for _, decl := range module.Decls {
+		dbDecl, ok := decl.(*schema.Database)
+		if !ok {
+			continue
+		}
+		b.dsns = append(b.dsns, dsnEntry{
+			name:   dbDecl.Name,
+			dbType: DBTypePostgres,
+			resolver: envarResolver{
+				name: fmt.Sprintf("FTL_POSTGRES_DSN_%s_%s", strings.ToUpper(module.Name), strings.ToUpper(dbDecl.Name)),
+			},
+		})
+	}
+	return b
+}
+
+func (b Builder) AddConfigFromManager(cm *cf.Manager[cf.Configuration]) Builder {
+	b.configs = append(b.configs, configManager[cf.Configuration]{manager: cm})
+	return b
+}
+
+func (b Builder) AddSecretsFromManager(sm *cf.Manager[cf.Secrets]) Builder {
+	b.secrets = append(b.secrets, configManager[cf.Secrets]{manager: sm})
+	return b
+}
+
+func refFromProto(r *ftlv1.ModuleContextResponse_Ref) ref {
+	return ref{
+		Module: optional.Ptr(r.Module),
+		Name:   r.Name,
+	}
+}
+
+type valueResolver struct {
+	value any
+}
+
+func (r valueResolver) Resolve() (isData bool, value any, data []byte, err error) {
+	return false, r.value, nil, nil
+}
+
+func (r valueResolver) ResolveString() (string, error) {
+	str, ok := r.value.(string)
+	if !ok {
+		return "", fmt.Errorf("value is not a string: %v", r.value)
+	}
+	return str, nil
+}
+
+type dataResolver struct {
+	data []byte
+}
+
+func (r dataResolver) Resolve() (isData bool, value any, data []byte, err error) {
+	return true, nil, r.data, nil
+}
+
+type envarResolver struct {
+	name string
+}
+
+func (r envarResolver) Resolve() (isData bool, value any, data []byte, err error) {
+	value, err = r.ResolveString()
+	if err != nil {
+		return false, nil, nil, err
+	}
+	return false, value, nil, nil
+}
+
+func (r envarResolver) ResolveString() (string, error) {
+	value, ok := os.LookupEnv(r.name)
+	if !ok {
+		return "", fmt.Errorf("missing environment variable %q", r.name)
+	}
+	return value, nil
+}
```

Comment:
The builder feels like a lot of machinery to do what seems like it should be fairly straightforward? Currently there's only a config manager, a secret manager, and DB envars. Can all this code just be a single function that takes those and outputs a ModuleContext?

Reply:
Moved this to just a helper function in the controller packager rather than using modulecontext.
It's in a separate file to avoid bloating controller.go more